### ISSUE TITLE
Algolia addObjects called with empty model data calls api

### ISSUE THREAD

**Issue by mnightingale** [laravel/scout#304]
If line 55 is executed we end up calling `$index->addObjects([])` which makes a request to Algolia.

https://github.com/laravel/scout/blob/d5b691339d707b465d337d8828641fd5c4fb706a/src/Engines/AlgoliaEngine.php#L49-L59

I'm not sure whether this behaviour should be improved in Scout or algolia/algoliasearch-client-php?

**Comment by julienbourdeau** [comment]
Hmm that's a good question, I would have said that it should be fixed in Scout but now that I think about it, maybe it makes sense to fix in the client so more people avoid making empty calls 🤔

**Comment by driesvints** [comment]
Is it though? Doesn't the `filter` method filter out empty models?

**Comment by driesvints** [comment]
Going to close this. Please provide a code example on how to reproduce this and I'll take another look.

**Comment by mnightingale** [comment]
**TL;DR:** If a models toSearchableArray() returns an empty array the current behaviour in the AlgoliaEngine basically ends up calling the equivalent to `collect([null])->filter()->values()->all()`which returns [] therefore we call `$index->addObjects([])` and a wasteful API request is made to Algolia, since the code filters out empty data I think the intended behaviour was not to call an indexing operation.

PR #318

----

I have a use case where 4,000 models get updated from an external feed several times per hour.
A model may have no changes but I still want to call `save()` so that the updated_at field gets touched.
Another situation is that not all of the fields that are updated are present in my index, therefore based on several comments in #153 I return `[]` in `toSearchableArray`.

I can't call `unsearchable()` or return false in `shouldBeSearchable()` because they'll both remove the model from the index which isn't what I want, I just don't want to reindex the model.

```php
public function toSearchableArray()
{
    if (!$this->isDirty('name')) {
        return []; // I don't want to reindex this model but I don't want to remove it
    }

    return [
        'name' => $this->name
    ];
}
```

The issue is that the update method of the engine will call `$index->addObjects([])` and still issue an empty update to Algolia wasting an API request without actually doing anything.

https://github.com/laravel/scout/blob/fa82a4d884b4c1b2d36e0eed4dfaccd6033662a1/src/Engines/AlgoliaEngine.php#L37-L60 

I think either it should check if the `$models->map(...)` call returns an empty array or the Algolia client should check if it receives an empty addObjects request.

Based on these lines, I think it's supposed to handle empty data itself?

https://github.com/laravel/scout/blob/fa82a4d884b4c1b2d36e0eed4dfaccd6033662a1/src/Engines/AlgoliaEngine.php#L54-L56
